### ISSUE TITLE
Problem with CR line separator?

### ISSUE THREAD

**Issue by mumrah** [uniVocity/univocity-parsers#8]
If I have a file with carriage returns as the line separator, like:

`a,b,c,d
1,2,3,4
5,6,7,8`

I end up with 1 row of 10 elements, rather than 3 rows of 4 (two get swallowed by the CR). For other test files, I get an EOFException. Here is a failing unit test:

```java
  public void testCarriageReturn() throws Exception {
    byte[] bytes = "a,b,c,d\r1,2,3,4\r5,6,7,8".getBytes(StandardCharsets.UTF_8);
    InputStream is = new ByteArrayInputStream(bytes);
    CsvParserSettings settings = new CsvParserSettings();
    settings.getFormat().setLineSeparator("\r");
    //settings.getFormat().setNormalizedNewline('\r');
    CsvParser parser = new CsvParser(settings);
    List<String[]> rows = parser.parseAll(new InputStreamReader(is));
    Assert.assertEquals(rows.size(), 3);
    Assert.assertEquals(rows.get(0).length, 4);
  }
```

The problem is alleviated by setting the normalizedNewline to the same value as the lineSeparator. This feels a little strange, and certainly seems contrary to what the documentation says. After reading http://docs.univocity.com/parsers/1.0.2/com/univocity/parsers/common/Format.html, I would think setting just lineSeparator to CR would be sufficient for this case. When would one need to specify normalizedNewline?

Am I misunderstanding the purpose of normalizedNewline and lineSeparator, or is this a bug? Or perhaps it's both?

Thanks!

**Comment by jbax** [comment]
If you run this code on the Windows platform, the JVM will convert the line separators automatically to match the platform's default.

So your input: "a,b,c,d\r1,2,3,4\r5,6,7,8" will become "a,b,c,d\r\n1,2,3,4\r\n5,6,7,8".

Try setting the lineSeparator to "\r\n" (I know it is weird) and let me know if it works for you.

**Comment by mumrah** [comment]
Well, in my case, I'm processing CSV data coming in from a JAX-RS resource as an InputStream rather than files. I wouldn't imagine that the platform would affect a stream of opaque bytes coming across the network. Since the CSV data coming in is user submitted, the format can be widely varying. The code will also be running on any number of different platforms. 

The functionality I'm hoping for is: user sends in CSV as a UTF-8 string, sets the line delimiter, and doesn't need to know or care about the server's platform (or newline setting)

What I have as a workaround at the moment is: 

```java
    String lineSeparator; // read from config
    settings.getFormat().setLineSeparator(lineSeparator);
    if(lineSeparator.length() == 1) {
      settings.getFormat().setNormalizedNewline(lineSeparator.charAt(0));
    }
```

which seems to work for the common cases, but is weird and probably incorrect in some cases. 

Also, per your suggestion, setting lineSeparator to `\r\n` with the default normalizedNewline did not work - I still need to set normalizedNewline to `\r`. I'm on OS X, if that matters.

I'm curious why the newline character needs to be considered at all? Univocity is dealing with a Reader which just returns bytes. In the parser, how is the lineSeparator treated differently from the newline?

**Comment by jbax** [comment]
uniVocity uses the Reader to read characters, and in this case the \n will be converted to \r\n automatically (on windows), even if the \r is not present in the input.

You should set the normalized newline character in the case of the \r line separator. What happens is: the parsers/writers read the normalized character. If you leave the default setting ('\n'), then no newlines will be detected. That's why you got all data in a single row (no new lines found)

The lineSeparator is used to convert sequences such as "\r\n" to a single character (the normalizedNewLine) when parsing. When writing, the normalizedNewLine will be converted to the lineSeparator you configured. 

You need this to process and generate files used by different platforms. For example, if you receive a file with \r\n on your macOS (e.g. "a,\r\nb", you don't want to get "a,\nb" but "a,b" instead).

**Comment by mumrah** [comment]
@jbax thanks, I think I understand things a little better now.

Taking a closer look at `AbstractCharInputReader`, I'm not sure it's doing exactly what you say. 

In the case that I have a single char line separator, like `\r`, it will never get replaced by the normalized newline due to this condition https://github.com/uniVocity/univocity-parsers/blob/master/src/main/java/com/univocity/parsers/common/input/AbstractCharInputReader.java#L136. In the constructor `lineSeparator2` is set to `\0` in the presence of a single-char separator.

If you try the unit test I posted up top and step through this method, you'll see what I mean.

**Comment by jbax** [comment]
You're right! Thanks for the details. Expect this issue to be resolved on version 1.3.2

**Comment by mumrah** [comment]
Thanks! Glad I'm not going crazy!

**Comment by jbax** [comment]
Fixed in commit https://github.com/uniVocity/univocity-parsers/commit/a158463382bd4a82298f0d03676d48e957347852

I've just released version 1.3.2. It may take some time to get the release available through maven, but you can get it directly from this link from the maven central:

http://oss.sonatype.org/content/repositories/releases/com/univocity/univocity-parsers/1.3.2/univocity-parsers-1.3.2.jar

**Comment by jbax** [comment]
And thank you very much for your contribution!
